lib/mpifx_shared_memory.fpp

```diff
@@ -0,0 +1,150 @@
+#:include 'mpifx.fypp'
+#:set TYPES = NUMERIC_TYPES
+
+!> Contains utilities for handling MPI shared memory
+module mpifx_shared_memory_module
+  use mpifx_common_module
+  use iso_c_binding, only : c_ptr, c_f_pointer
+  implicit none
+  private
+
+  public :: mpifx_allocate_shared, mpifx_free_shared, mpifx_lock_shared, mpifx_unlock_shared, mpifx_sync_shared
+
+  interface mpifx_allocate_shared
+#:for TYPE in TYPES
+  #:set TYPEABBREV = TYPE_ABBREVS[TYPE]
+  module procedure mpifx_allocate_shared_${TYPEABBREV}$
+#:endfor
+  end interface mpifx_allocate_shared
+
+contains
+
+#:def mpifx_allocate_shared_template(SUFFIX, TYPE)
+
+  !> Returns a window handle and a pointer to the address associated with a shared memory segment.
+  !!
+  !! \param mycomm  MPI communicator.
+  !! \param length  Number of elements of type ${TYPE}$ in the shared memory window.
+  !! \param win  Handle of the shared memory window on return.
+  !! \param shared_data  Pointer to the shared data array of length 'length' on return.
+  !! \param error  Optional error code on return.
+  !!
+  !! \see MPI documentation (\c MPI_WIN_ALLOCATE_SHARED)
+  !!
+  subroutine mpifx_allocate_shared_${SUFFIX}$(mycomm, length, win, shared_data, error)
+    type(mpifx_comm), intent(in) :: mycomm
+    integer, intent(in) :: length
+    integer, intent(out) :: win
+    ${TYPE}$, pointer, intent(out) :: shared_data(:)
+    integer, intent(out), optional :: error
+
+    integer :: disp_unit, error0, error1
+    integer(MPI_ADDRESS_KIND) :: local_length
+    type(c_ptr) :: baseptr
+
+    disp_unit = kind(shared_data)
```

Comment:
```suggestion
    disp_unit = storage_size(shared_data) / 8
```
Note: As per Fortran standard, `kind()` returns an arbitrary positive integer number used to enumerate the kinds of a given variable type. For some compilers (intel, gnu) it is accidentally the same, as the number of bytes the given kind occupies, but for some other compilers it is not (e.g. nag returns `1` for single precision and `2` for double precision!).  We should use `storage_size()` instead, which returns the nr. of bits an element of a given type within an array would occupy (including eventual padding in theory).

Reply:
Thanks for pointing this out. I have changed that. :+1: